sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala

```diff
@@ -177,4 +120,83 @@ object RewriteWithExpression extends Rule[LogicalPlan] {
       case other => other.mapChildren(rewriteWithExprAndInputPlans(_, inputPlans))
     }
   }
+
+  private def genProject(
+      defs: Seq[Expression],
+      inputPlans: Array[LogicalPlan]): mutable.Map[CommonExpressionId, Expression] = {
+    val refToExpr = mutable.HashMap.empty[CommonExpressionId, Expression]
+    val childProjections = Array.fill(inputPlans.length)(mutable.ArrayBuffer.empty[Alias])
+    defs.zipWithIndex.foreach { case (CommonExpressionDef(child, id), index) =>
+      if (id.canonicalized) {
+        throw SparkException.internalError(
+          "Cannot rewrite canonicalized Common expression definitions")
+      }
+
+      if (CollapseProject.isCheap(child)) {
+        refToExpr(id) = child
+      } else {
```

Comment:
This seems potentially less than ideal, it looks like CollapseProject.isCheap will return true for Python UDFS. It also _might_ violate the promise of being gauranteed to be evaluated only once (see With.scala L26).